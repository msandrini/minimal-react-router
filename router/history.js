/* global window */

const historyListeners = [];

window.onpopstate = (event) => {
	historyListeners.forEach(callback => callback(event.path[0].location.pathname));
};

export const push = (pathname) => {
	window.history.pushState({}, '', pathname);
	historyListeners.forEach(callback => callback(pathname));
};

export const onChangeLocation = (callback) => {
	historyListeners.push(callback);
};

window.hLTemp = historyListeners;
