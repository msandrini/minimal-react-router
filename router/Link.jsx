import React from 'react';
import PropTypes from 'prop-types';

import * as history from './history';

const Link = (props) => {
	const onClick = (e) => {
		const aNewTab = e.metaKey || e.ctrlKey;
		// TODO: support other protocols with regex
		const anExternalLink = props.to.startsWith('http');

		if (!aNewTab && !anExternalLink) {
			e.preventDefault();
			history.push(props.to);
		}
	};

	const activeClass = props.isActive ? props.activeClassName : '';
	const className = `${props.className}${activeClass}`;

	return (
		<a
			{...props}
			href={props.to}
			className={className}
			onClick={onClick}
		>
			{props.children}
		</a>
	);
};

Link.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.string
	]),
	className: PropTypes.string,
	to: PropTypes.string.isRequired,
	activeClassName: PropTypes.string,
	isActive: PropTypes.bool
};

Link.defaultProps = {
	children: null,
	className: '',
	activeClassName: 'is-active',
	isActive: false
};

export default Link;
