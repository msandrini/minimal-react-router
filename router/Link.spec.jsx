import React from 'react';
import { shallow } from 'enzyme';

import Link from './Link';

import * as history from './history';

describe('Link', () => {
	it('should render a link navigation based on props', () => {
		const wrapper = shallow(<Link to="/home" />);
		expect(wrapper).toMatchSnapshot();
	});
	it('should have active class is active', () => {
		const wrapper = shallow(<Link to="/home" isActive />);
		expect(wrapper).toMatchSnapshot();
	});
	describe('onClick', () => {
		it('should use local history if local path', () => {
			history.push = jest.fn();

			const wrapper = shallow(<Link to="/home" />);
			let link = wrapper.find('a');

			link.simulate('click', { preventDefault: jest.fn() });
			expect(history.push).toHaveBeenCalledTimes(1);

			const wrapper2 = shallow(<Link to="http://www.pudim.com.br" />);
			link = wrapper2.find('a');

			link.simulate('click', { preventDefault: jest.fn() });
			expect(history.push).toHaveBeenCalledTimes(1);
		});
	});
});
