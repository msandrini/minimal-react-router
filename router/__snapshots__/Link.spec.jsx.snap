// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should have active class is active 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Link
    activeClassName="is-active"
    className=""
    isActive={true}
    to="/home"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": null,
      "className": "is-active",
      "href": "/home",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": "a",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": null,
        "className": "is-active",
        "href": "/home",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": "a",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Link should render a link navigation based on props 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Link
    activeClassName="is-active"
    className=""
    isActive={false}
    to="/home"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": null,
      "className": "",
      "href": "/home",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": "a",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": null,
        "className": "",
        "href": "/home",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": "a",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
